include/deal.II/sundials/ida_interface.h

```diff
@@ -0,0 +1,476 @@
+     * set Maximum order of BDF                          = 5
+     * set Min step size                                 = 0.000001
+     * set Relative error tolerance                      = 0.000010
+     * set Show output of time steps                     = true
+     * set Time units between each output                = 0.05
+     * set Use local tolerances                          = false
+     * @endcode
+     *
+     * These are one-to-one with the options you can pass at construction time.
+     *
+     * Those options are set as default value in the ParameterHandler object
+     * `prm`.
+     */
+    virtual void add_parameters(ParameterHandler &prm);
+
+    /**
+     * Evolve. This function returns the final number of steps.
+     */
+    unsigned int solve_dae(VectorType &solution,
+                           VectorType &solution_dot);
+
+    /**
+     * Clear internal memory, and start with clean objects. This function is
+     * called when the simulation start and when the user return true to a call
+     * to solver_should_restart()
+     */
+    void reset_dae(const double t,
+                   VectorType &y,
+                   VectorType &yp,
+                   double h,
+                   bool first_step);
+
+    /**
+     * Return a newly allocated shared_ptr<VectorType>.
+     */
+    std::function<std::shared_ptr<VectorType>()> create_new_vector;
+
+    /**
+     * Compute residual.
+     */
+    std::function<int(const double t,
+                      const VectorType &y,
+                      const VectorType &y_dot,
+                      VectorType &res)> residual;
+
+    /**
+     * Compute Jacobian.
+     */
+    std::function<int(const double t,
+                      const VectorType &y,
+                      const VectorType &y_dot,
+                      const double alpha)> setup_jacobian;
+
+    /**
+     * Solve linear system.
+     */
+    std::function<int(const VectorType &rhs, VectorType &dst)> solve_jacobian_system;
+
+    /**
+     * Store solutions to file.
+     */
+    std::function<void (const double t,
+                        const VectorType &sol,
+                        const VectorType &sol_dot,
+                        const unsigned int step_number)> output_step;
+
+    /**
+     * Evaluate wether the solver should be restarted (for example because the
+     * number of degrees of freedom has changed).
+     *
+     * This function is supposed to perform all operations that are necessary in
+     * `sol` and `sol_dot` to make sure that the resulting vectors are
+     * consistent, and of the correct final size.
+     */
+    std::function<bool (const double t,
+                        VectorType &sol,
+                        VectorType &sol_dot)> solver_should_restart;
+
+    /**
+     * Return a vector whose component are 1 if the corresponding
+     * dof is differential, 0 if algebraic.
+     */
+    std::function<VectorType&()> differential_components;
```

Comment:
If you only need zeros or ones, why not let the function return a `std::vector<bool>`? The fact that *internally* IDE wants a different format shouldn't dictate using a better interface as we can always do the conversion internally.

Reply:
Maybe even better: we could return an IndexSet.

Reply:
Yes, definitely better!